src/core/friendly_errors/param_validator.js

```diff
@@ -8,351 +8,366 @@ import { z } from 'zod';
 import dataDoc from '../../../docs/parameterData.json';
 
-// Cache for Zod schemas
-let schemaRegistry = new Map();
-const arrDoc = JSON.parse(JSON.stringify(dataDoc));
-
-// Mapping names of p5 types to their constructor functions.
-// p5Constructors:
-//   - Color: f()
-//   - Graphics: f()
-//   - Vector: f()
-// and so on.
-const p5Constructors = {};
-// For speedup over many runs. `funcSpecificConstructors[func]` only has the
-// constructors for types which were seen earlier as args of `func`.
-// const funcSpecificConstructors = {};
-
-for (let [key, value] of Object.entries(p5)) {
-  p5Constructors[key] = value;
-}
-
-// window.addEventListener('load', () => {
-//   // Make a list of all p5 classes to be used for argument validation
-//   // This must be done only when everything has loaded otherwise we get
-//   // an empty array.
-//   for (let key of Object.keys(p5)) {
-//     // Get a list of all constructors in p5. They are functions whose names
-//     // start with a capital letter.
-//     if (typeof p5[key] === 'function' && key[0] !== key[0].toLowerCase()) {
-//       p5Constructors[key] = p5[key];
-//     }
-//   }
-// });
-
-// `constantsMap` maps constants to their values, e.g.
-// {
-//   ADD: 'lighter',
-//   ALT: 18,
-//   ARROW: 'default',
-//   AUTO: 'auto',
-//   ...
-// }
-const constantsMap = {};
-for (const [key, value] of Object.entries(constants)) {
-  constantsMap[key] = value;
-}
-
-const schemaMap = {
-  'Any': z.any(),
-  'Array': z.array(z.any()),
-  'Boolean': z.boolean(),
-  'Function': z.function(),
-  'Integer': z.number().int(),
-  'Number': z.number(),
-  'Number[]': z.array(z.number()),
-  'Object': z.object({}),
-  // Allows string for any regex
-  'RegExp': z.string(),
-  'String': z.string(),
-  'String[]': z.array(z.string())
-};
-
-// const webAPIObjects = [
-//   'AudioNode',
-//   'HTMLCanvasElement',
-//   'HTMLElement',
-//   'KeyboardEvent',
-//   'MouseEvent',
-//   'TouchEvent',
-//   'UIEvent',
-//   'WheelEvent'
-// ];
-
-// function generateWebAPISchemas(apiObjects) {
-//   return apiObjects.map(obj => {
-//     return {
-//       name: obj,
-//       schema: z.custom((data) => data instanceof globalThis[obj], {
-//         message: `Expected a ${obj}`
-//       })
-//     };
-//   });
-// }
-
-// const webAPISchemas = generateWebAPISchemas(webAPIObjects);
-
-/**
- * This is a helper function that generates Zod schemas for a function based on
- * the parameter data from `docs/parameterData.json`.
- *
- * Example parameter data for function `background`:
- * "background": {
-      "overloads": [
-        ["p5.Color"],
-        ["String", "Number?"],
-        ["Number", "Number?"],
-        ["Number", "Number", "Number", "Number?"],
-        ["Number[]"],
-        ["p5.Image", "Number?"]
-      ]
+function validateParams(p5, fn) {
+  // Cache for Zod schemas
+  let schemaRegistry = new Map();
+  const arrDoc = JSON.parse(JSON.stringify(dataDoc));
```

Comment:
Is this duplication because we modify the entries?

Reply:
What do you mean by duplication here?

Reply:
the JSON.parse(JSON.stringify(...)) bit, normally I see that used to make a deep copy of an object

Reply:
If there is a need for duplicating object use [`structuredClone()`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone)

Reply:
We also might not to do a clone at all here if we're only reading the data and not modifying it later

Reply:
Ahh yes! I actually got this line of code from the old parameter validation and didn't think too much about it. Changed the code so that it uses `dataDoc` directly.